src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/Bau/DataSetFileMetaMigrationController.cs

```diff
@@ -0,0 +1,136 @@
+#nullable enable
+using GovUk.Education.ExploreEducationStatistics.Content.Model.Database;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using GovUk.Education.ExploreEducationStatistics.Common.Extensions;
+using GovUk.Education.ExploreEducationStatistics.Common.Model;
+using GovUk.Education.ExploreEducationStatistics.Common.Model.Data;
+using GovUk.Education.ExploreEducationStatistics.Content.Model;
+using GovUk.Education.ExploreEducationStatistics.Data.Model.Database;
+using Microsoft.EntityFrameworkCore;
+using static GovUk.Education.ExploreEducationStatistics.Admin.Models.GlobalRoles;
+
+namespace GovUk.Education.ExploreEducationStatistics.Admin.Controllers.Api.Bau;
+
+[Route("api")]
+[ApiController]
+[Authorize(Roles = RoleNames.BauUser)]
+public class DataSetFileMetaMigrationController : ControllerBase
+{
+    private readonly ContentDbContext _contentDbContext;
+    private readonly StatisticsDbContext _statisticsDbContext;
+
+    public DataSetFileMetaMigrationController(
+        ContentDbContext contentDbContext,
+        StatisticsDbContext statisticsDbContext)
+    {
+        _contentDbContext = contentDbContext;
+        _statisticsDbContext = statisticsDbContext;
+    }
+
+    public class DataSetFileMetaMigrationResult
+    {
+        public bool IsDryRun;
+        public int Processed;
+        public int LeftToProcess;
+    }
+
+    public class Observation
+    {
+        public GeographicLevel GeographicLevel;
+        public int Year;
+        public TimeIdentifier TimeIdentifier;
+    }
+
+    [HttpPatch("bau/migrate-datasetfilemeta")]
+    public async Task<DataSetFileMetaMigrationResult> MigrateReleaseSeries(
+        [FromQuery] bool dryRun = true,
+        [FromQuery] int? num = null,
+        CancellationToken cancellationToken = default)
+    {
+        var filesQueryable = _contentDbContext.Files
+            .Where(f =>
+                f.DataSetFileMeta == null
+                && f.Type == FileType.Data);
+
+        if (num is > 0)
+        {
+            filesQueryable = filesQueryable.Take(num.Value);
+        }
+
+        var files = await filesQueryable.ToListAsync(cancellationToken: cancellationToken);
+
+        var numDataSetFileMetaSet = 0;
+
+        foreach (var file in files)
+        {
+            var observations = await _statisticsDbContext.Observation
```

Comment:
Another update: 

I actually started to wonder if the original was actually faster. My reasoning was that with two queries, the DB would have to do two sweeps over the observations instead of one.

So I tested them both locally and (to my surprise) the two queries was significantly faster, so going to go with that.

Reply:
Cool - nice one on doing the timing testing on this :+1: